Refactor & test fast beam search method, add ngram blocking

flauted opened the pull request:
This builds off of #1256 . Either this fixes that (and that should be closed) or that should be merged first.

This PR does a few things (separate from 1256):
1) Refactors translator.py ``_fast_translate_batch`` to use a beam object. This makes it much easier to compare ``_fast_translate_batch`` and ``_translate_batch`` line-by-line.
1.5) Tests that object similar to how 1256 tests the ``Beam`` object.
2) In translator.py ``translate``, uses the new ``_log()`` function in places that were basically in-lining it.
3) In beam.py ``Beam`` changes ``get_current_origin()``, ``get_current_state()``, and ``done()`` into idiomatic properties.
4) In translation.py, ``Translation`` instead of constructing the output string with ``+=``, uses ``"".join()`` (see, e.g., [here](https://stackoverflow.com/a/3055541/8217456) for why that's better) and uses ``__slots__`` for properties since lots and lots of of ``Translation``s are instantiated (faster attribute access, less memory used - see, .e.g., [here](https://stackoverflow.com/a/28059785/8217456)).
5) Addresses ``data`` arguments all over translator.py. The docstring of ``translate_batch()`` currently has a note about how that's unnecessary.

---

Speed/accuracy tests using the pretrained transformer (averaged-10-epoch.pt from the website) and the ``head -n 500`` of ``test.en`` and ``test.de``.

**Before:**
Slow:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time

PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 40.353486
Average translation time (s): 0.080707
Tokens per second: 299.230652
```
Fast:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -fast

PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 9.664046
Average translation time (s): 0.019328
Tokens per second: 1249.476631
```

**After:**
Slow:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time

PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 40.739757
Average translation time (s): 0.081480
Tokens per second: 296.393522
```
Fast:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -fast

PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 9.451516
Average translation time (s): 0.018903
Tokens per second: 1277.572869
```
I don't have a great explanation for why this is faster. Slots and join in ``Translation`` should give a slight boost, but testing that alone doesn't give an appreciable speed difference.

vince62s commented:
@bpopeters  you may comment as well since this may make #862 obsolete.

bpopeters commented:
I wouldn't worry about #862. If it were going to get merged, the time to do so was in July. At this point, the conflicts to fix are probably more work than I put into the actual contributions.

As for this pull request, is it a step that will eventually result in there being one implementation of translation in `translator.py` instead of three? That, to me, is a pressing issue.

guillaumekln commented on onmt/translate/translator.py:
```diff
@@ -562,184 +548,81 @@ def _fast_translate_batch(
+            select_indices = beam.current_origin
 
             # Reorder states.
             if isinstance(memory_bank, tuple):
```
I wanted to make this change some time ago: source side vectors (here `memory_bank`, `memory_lengths`, and `src_map`) do not need to be reordered at each step, only when some batches are marked for removal.

Maybe this could be moved under `if beam.is_finished.any():`?

flauted replied:
Done. Oddly seems to be slower,
```
PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 9.625983
Average translation time (s): 0.019252
Tokens per second: 1254.417403
```
but still a good change. (I think it's expected, but the decoder state needs mapped every step or the score changes)

vince62s commented on onmt/tests/test_beam.py:
```diff
@@ -0,0 +1,371 @@
+
+class TestBeam(unittest.TestCase):
+    BLOCKED_SCORE = -10e20
+
```
here -10e20
later -1e20
do we need to switch to stuff like -inf for future fp16 support or it's a none issue ?

guillaumekln replied:
The scores should always be in FP32.

vince62s replied:
ok. but we have a bit of everything in the code like here https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/multi_headed_attn.py#L172
just a comment, if you think it's ok, no big deal.

flauted commented on onmt/translate/translator.py:
```diff
@@ -562,184 +548,81 @@ def _fast_translate_batch(
         assert not self.dump_beam
         assert not self.use_filter_pred
-        assert self.block_ngram_repeat == 0
+        # assert self.block_ngram_repeat == 0
```
Oops. It's not. I was working on that when I decided I need tests.

flauted commented on onmt/translate/fast_beam.py:
```diff
@@ -0,0 +1,173 @@
+import torch
+
+
+class FastBeamBatch(object):
```
Done. What about the file name?

guillaumekln replied:
The existing `beam.py` makes it a bit confusing, otherwise I would simply go for `beam_search.py`. @bpopeters @vince62s Any recommendation here?

vince62s replied:
beam_search.py is fine and allows us to merge with the regular beam later.

vince62s commented:
@flauted  can you please change in your PR the travis.yml file and swithc to Python 3.6 instead ? (it is in several lines) not sure 100% this is the cause of travsi failing, but can't hurt.

flauted commented:
@vince62s I can but I don't understand. Travis didn't fail.

vince62s commented:
no you right, after a restart it went ok.

vince62s commented:
@guillaumekln it seems ok to merge, right ?

flauted commented:
I added ngram blocking to ``BeamSearch``. Basically uses the same implementation from the slow ``Beam`` (and it's still >10x faster). The scores are exactly the same as the slow code. Also I added some attributes to ``Translator`` so that they don't need calculated every time ``translate_batch`` is called. This gives a slight speed improvement.

Some of the tests also had small "bugs" before where they weren't testing what they should have been. Last commit cleans that up.

-----
## Scores and Times

**Block repeat with ignores**
Slow:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -block_ngram_repeat 3 -ignore_when_blocking ▁der ▁in

PRED AVG SCORE: -0.4415, PRED PPL: 1.5551
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 290.394258
Average translation time (s): 0.580789
Tokens per second: 41.581401
```
Fast:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -block_ngram_repeat 3 -ignore_when_blocking ▁der ▁in -fast

PRED AVG SCORE: -0.4415, PRED PPL: 1.5551
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 23.889659
Average translation time (s): 0.047779
Tokens per second: 505.448824
```

**Block repeat, no ignores**
Slow:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -block_ngram_repeat 3

PRED AVG SCORE: -0.4417, PRED PPL: 1.5554
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 292.230435
Average translation time (s): 0.584461
Tokens per second: 41.309866
```
Fast:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -block_ngram_repeat 3 -fast

PRED AVG SCORE: -0.4417, PRED PPL: 1.5554
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 24.011137
Average translation time (s): 0.048022
Tokens per second: 502.766695
```

**No blocking** (Matches both the before and after from above)
Slow:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time

PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 39.648572
Average translation time (s): 0.079297
Tokens per second: 304.550688
```
Fast:
```
python translate.py -model /home/dylan/Downloads/averaged-10-epoch.pt -src /home/dylan/Downloads/test_short.en -tgt /home/dylan/Downloads/test_short.de -verbose -replace_unk -output pred.txt -gpu 0 -report_time -fast

PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 9.489469
Average translation time (s): 0.018979
Tokens per second: 1272.463152
```

guillaumekln commented:
Yes, I think we can merge this. Next step could be to better isolate some logic like ngram blocking, penalties to make the code simpler and easier to extend. And then we should be very close to removing the original decoding code.
